internal/app/machined/pkg/controllers/network/address_status.go

```diff
@@ -0,0 +1,175 @@
+)
+
+// AddressStatusController manages secrets.Etcd based on configuration.
+type AddressStatusController struct{}
+
+// Name implements controller.Controller interface.
+func (ctrl *AddressStatusController) Name() string {
+	return "network.AddressStatusController"
+}
+
+// Inputs implements controller.Controller interface.
+func (ctrl *AddressStatusController) Inputs() []controller.Input {
+	return nil
+}
+
+// Outputs implements controller.Controller interface.
+func (ctrl *AddressStatusController) Outputs() []controller.Output {
+	return []controller.Output{
+		{
+			Type: network.AddressStatusType,
+			Kind: controller.OutputExclusive,
+		},
+	}
+}
+
+// Run implements controller.Controller interface.
+//
+//nolint:gocyclo
+func (ctrl *AddressStatusController) Run(ctx context.Context, r controller.Runtime, logger *log.Logger) error {
+	watchConn, err := rtnetlink.Dial(&netlink.Config{
+		Groups: unix.RTMGRP_LINK | unix.RTMGRP_IPV4_IFADDR | unix.RTMGRP_IPV6_IFADDR,
+	})
+	if err != nil {
+		return fmt.Errorf("error dialing watch socket: %w", err)
+	}
+
+	var wg sync.WaitGroup
+
+	wg.Add(1)
+
+	watchCh := make(chan struct{})
+
+	go func() {
+		defer wg.Done()
+
+		for {
+			_, _, watchErr := watchConn.Receive()
+			if watchErr != nil {
+				return
+			}
+
+			select {
+			case watchCh <- struct{}{}:
+			case <-ctx.Done():
+				return
+			}
+		}
+	}()
+
+	defer wg.Wait()
+
+	// close the watchConn first to abort the goroutine above on early exit
+	defer watchConn.Close() //nolint:errcheck
+
+	conn, err := rtnetlink.Dial(nil)
+	if err != nil {
+		return fmt.Errorf("error dialing rtnetlink socket: %w", err)
+	}
+
+	defer conn.Close() //nolint:errcheck
+
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case <-r.EventCh():
+		case <-watchCh:
+		}
+
+		// build links lookup table
+		links, err := conn.Link.List()
+		if err != nil {
+			return fmt.Errorf("error listing links: %w", err)
+		}
+
+		linkLookup := make(map[uint32]string, len(links))
+
+		for _, link := range links {
+			linkLookup[link.Index] = link.Attributes.Name
+		}
+
+		// list resources for cleanup
+		list, err := r.List(ctx, resource.NewMetadata(network.NamespaceName, network.AddressStatusType, "", resource.VersionUndefined))
+		if err != nil {
+			return fmt.Errorf("error listing resources: %w", err)
+		}
+
+		itemsToDelete := map[resource.ID]struct{}{}
+
+		for _, r := range list.Items {
+			itemsToDelete[r.Metadata().ID()] = struct{}{}
+		}
+
+		addrs, err := conn.Address.List()
+		if err != nil {
+			return fmt.Errorf("error listing addresses: %w", err)
+		}
+
+		for _, addr := range addrs {
+			addr := addr
+
+			// TODO: should we use local address actually?
+			// from if_addr.h:
+			// IFA_ADDRESS is prefix address, rather than local interface address.
+			// * It makes no difference for normally configured broadcast interfaces,
+			// * but for point-to-point IFA_ADDRESS is DESTINATION address,
+			// * local address is supplied in IFA_LOCAL attribute.
```

Comment:
Maybe have a special handler for p2p interfaces?  Definitely an important TODO, and p2p interfaces are fairly common in multi-tenant DCs.  I'm pretty sure Nokia use them, for instance.

I definitely prefer having the prefix information (subnet mask), so I'd hate to lose that in the general case.

Reply:
On second though, I think the way you implement it (with `Local`) is fine.  It feels weird, but it follows the kernel's implementation.